improve vectorchord build index and search option (#113)
tensorchord/pgvecto.rs-docs: src/vectorchord/getting-started/vectorchord-suite.md

What if you could build a powerful, production-ready RAG system directly within your trusted PostgreSQL database?

Enter the **VectorChord Suite**, a collection of PostgreSQL extensions designed to bring high-performance vector search, native BM25 ranking, and flexible tokenization capabilities right into Postgres. Let's explore the components and how they enable next-generation RAG solutions.

## What is the VectorChord Suite?

The suite comprises three key PostgreSQL extensions working in concert:

1. [**VectorChord:**](https://github.com/tensorchord/VectorChord) The core vector search engine. It's specifically designed for scalable, high-performance, and disk-efficient vector similarity search within PostgreSQL.
    
2. [**VectorChord-bm25:**](https://github.com/tensorchord/VectorChord-bm25) This extension implements the sophisticated BM25 ranking algorithm directly inside PostgreSQL, leveraging efficient Block-WeakAnd algorithms. BM25 is a standard for relevance ranking based on keyword frequency and document characteristics.
    

> 3. [**pg\_**](https://github.com/tensorchord/pg_tokenizer.rs)[**tokenizer.rs**](http://tokenizer.rs)[**:**](https://github.com/tensorchord/pg_tokenizer.rs) Provides essential text tokenization capabilities needed for effective full-text search, enabling fine-grained control over how text is processed for full-text search.

Author: fixed.

    

By combining these extensions, you unlock powerful capabilities for building advanced RAG systems entirely within PostgreSQL.

## How to Use the VectorChord Suite

> You can use the `tensorchord/vchord-suite` image to run multiple extensions which are provided by TensorChord. The image is based on the official Postgres image and includes the following extensions:

Reviewer: Includes what?

vector       | 0.8.0   | public            | vector data type and ivfflat and hnsw access methods
```

## Use Cases for the VectorChord Suite

### Use Case 1: Powerful Hybrid Search with Native BM25 and VectorChord

<img src="https://cdn.hashnode.com/res/hashnode/image/upload/v1744643437495/b57592c9-e1c2-4940-ae59-751ff93e37a3.jpeg" alt="Hybrid Search" width="600" align="center"/>

In the RAG era, effective retrieval is paramount. Neither keyword search nor vector search alone is perfect:

* **Keyword Search (like BM25):** Excels at precision, finding documents with exact keyword matches. It's great for structured queries and term-specific searches. However, it struggles with synonyms, paraphrasing, and understanding the underlying *meaning* or semantic intent. (Leverages `VectorChord-bm25` and `pg_tokenizer`).
    
* **Vector Search:** Captures deep semantic meaning and relationships between concepts, allowing it to find relevant information even if the exact keywords aren't present. However, it can sometimes lack precision for queries demanding specific term matches. (Leverages `VectorChord`).
    

> **The Solution: Hybrid Search.** By combining the strengths of both approaches within Postgres, the VectorChord Suite bridges this gap. You can run a query that leverages:

Reviewer: ```suggestion
**The Solution: Hybrid Search.** By combining the strengths of both approaches within PostgreSQL, the VectorChord Suite bridges this gap. You can run a query that leverages:
```

### Use Case 2: Beyond Text - OCR-Free RAG with ColQwen2 & VectorChord

Building RAG systems for documents like PDFs or scanned images often involves cumbersome pre-processing pipelines. Traditional methods rely heavily on Optical Character Recognition (OCR) and layout analysis to extract text. This process can be:

* **Slow:** OCR can be computationally expensive.
    
* **Error-Prone:** OCR accuracy varies significantly depending on document quality.
    
* **Lossy:** Crucial visual contexts like tables, figures, formatting, and relative positioning is often lost during text extraction.
    

**The Solution: OCR-Free, Visually-Aware RAG.** What if you could query documents based on *how they look* and their *content*, without explicit OCR?

This is now achievable by combining:

1. **Multi-Modal Vision Language Models (VLMs):** Models like ColQwen2 can process images (document pages) and generate embeddings that capture *both* textual content and visual layout information.
    

> 2. **VectorChord's Multi-Vector Capabilities:** VectorChord can efficiently store and search *multiple* vectors per document within Postgres – allowing you to store embeddings representing different aspects (e.g., text content, visual layout).

Reviewer: ```suggestion
2. **VectorChord's Multi-Vector Capabilities:** VectorChord can efficiently store and search *multiple* vectors per document within PostgreSQL – allowing you to store embeddings representing different aspects (e.g., text content, visual layout).
```

    

With this setup, you can query your document database using prompts that reference visual elements ("Find documents with a bar chart comparing sales figures") or combined textual and visual cues, directly within PostgreSQL. This simplifies your RAG stack, potentially boosts retrieval accuracy by preserving visual context, and eliminates the bottlenecks associated with traditional OCR pipelines.

## Why Build RAG in PostgreSQL?

> Leveraging the VectorChord Suite within Postgres offers significant advantages:

Reviewer: ```suggestion
Leveraging the VectorChord Suite within PostgreSQL offers significant advantages:
```

* **Unified Data:** Keep your source data, text, metadata, and vector embeddings all in one place.
    
* **Reduced Complexity:** Eliminate the need for separate vector databases and synchronization pipelines.
    

> * **Leverage Existing Infrastructure:** Utilize your existing Postgres expertise, tooling, and operational practices.

Reviewer: ```suggestion
* **Leverage Existing Infrastructure:** Utilize your existing PostgreSQL expertise, tooling, and operational practices.
```

    
* **Transactional Integrity:** Benefit from PostgreSQL's robust ACID compliance.
    

> * **Rich Ecosystem:** Access the wide array of tools and features available within the Postgres ecosystem.

Reviewer: ```suggestion
* **Rich Ecosystem:** Access the wide array of tools and features available within the PostgreSQL ecosystem.
```

    

## Conclusion

The VectorChord Suite transforms PostgreSQL into a powerful, production-ready platform for building advanced RAG solutions. By integrating high-performance vector search (`VectorChord`), sophisticated keyword ranking (`VectorChord-bm25`), and flexible text processing (`pg_tokenizer`), you can implement cutting-edge techniques like hybrid search and OCR-free multi-modal retrieval directly within your database. Simplify your architecture, enhance retrieval accuracy, and unlock the full potential of RAG with VectorChord in PostgreSQL.

> ## Previous Blog Post

Reviewer: ```suggestion
## Blog Posts
```

> In our previous blog, we shared some user cases where all images can be replaced with VectorChord Suite image.

Reviewer: ```suggestion
In our recent blog posts, we shared some user cases where all Docker images can be replaced with VectorChord Suite image.
```

    
2. [**VectorChord-bm25:**](https://github.com/tensorchord/VectorChord-bm25) This extension implements the sophisticated BM25 ranking algorithm directly inside PostgreSQL, leveraging efficient Block-WeakAnd algorithms. BM25 is a standard for relevance ranking based on keyword frequency and document characteristics.
    
3. [**pg_tokenizer.rs:**](https://github.com/tensorchord/pg_tokenizer.rs) Provides essential text tokenization capabilities needed for effective full-text search, enabling fine-grained control over how text is processed for full-text search.
    

By combining these extensions, you unlock powerful capabilities for building advanced RAG systems entirely within PostgreSQL.

## How to Use the VectorChord Suite

You can use the `tensorchord/vchord-suite` Docker image to run multiple extensions which are provided by TensorChord. The image is based on the Docker official PostgreSQL image and includes the following extensions:
- `vchord`
- `pg_tokenizer` 

> - `vchord_bm25`

Reviewer: and `vector`

> ```powershell

Reviewer: Is docker command supported by powershell?

Author: change to shell

> docker run   \           

Reviewer: ```suggestion
docker run \
```

  --name vchord-suite  \
  -e POSTGRES_PASSWORD=postgres  \
  -p 5432:5432 \
  -d tensorchord/vchord-suite:pg17-latest
  # If you want to use ghcr image, you can change the image to `ghcr.io/tensorchord/vchord-suite:pg17-latest`.

>   # if you want to use the specific version, you can use the tag `pg17-20250414`, supported version can be found in the support matrix.

Reviewer: Better if the command is given.

Author: fixed.

    

By combining these extensions, you unlock powerful capabilities for building advanced RAG systems entirely within PostgreSQL.

## How to Use the VectorChord Suite

You can use the `tensorchord/vchord-suite` Docker image to run multiple extensions which are provided by TensorChord. The image is based on the Docker official PostgreSQL image and includes the following extensions:
- `vchord`
- `pg_tokenizer` 
- `vchord_bm25`
- `vector`

```shell
docker run   \           
  --name vchord-suite  \
  -e POSTGRES_PASSWORD=postgres  \
  -p 5432:5432 \
  -d tensorchord/vchord-suite:pg17-latest

# If you want to use ghcr image, you can change the image to `ghcr.io/tensorchord/vchord-suite:pg17-latest`.
docker run   \           
  --name vchord-suite  \
  -e POSTGRES_PASSWORD=postgres  \
  -p 5432:5432 \
  -d tensorchord/vchord-suite:pg17-latest
  
# If you want to use the specific version, you can use the tag `pg17-20250414`, supported version can be found in the support matrix.
docker run   \           
  --name vchord-suite  \
  -e POSTGRES_PASSWORD=postgres  \
  -p 5432:5432 \
  -d tensorchord/vchord-suite:pg17-20250414

> ```

Reviewer: Remove tailing spaces.

Author: done
